crates/node/src/scheduler/mod.rs

```diff
@@ -205,7 +205,15 @@ impl Scheduler {
             // Before scheduling new workload, try to start pending programs first.
             {
                 let mut state = self.state.lock().await;
-                while let Some((tx_hash, program_id)) = state.pending_programs.pop_front() {
+
+                // Make a local copy of pending programs so that when some
+                // programs fail to start and are rescheduled, they won't
+                // cause an infinite loop here.
+                let mut current_pending_programs = state.pending_programs.clone();
+                while let Some((tx_hash, program_id)) = current_pending_programs.pop_front() {
+                    // Pop from the actual pending program queue as well.
+                    state.pending_programs.pop_front();
```

Comment:
This should be correct, right?

The original queue is cloned, so that on this scheduler round only "old" entries are processed.

If some of the programs fail to start and needs rescheduling, it's added to the end of the `state.pending_programs` so that it gets processed on the next full round of scheduler loop.